java-squid/src/main/java/org/sonar/java/se/AbstractSymbolicValueRelation.java

```diff
@@ -0,0 +1,166 @@
+    if (usedRelations.size() > 200) {
+      throw new IllegalStateException();
+    }
+    for (AbstractSymbolicValueRelation relation : knownRelations) {
+      Boolean result = relation.implies(this);
+      if (result != null) {
+        return result;
+      }
+      usedRelations.add(relation);
+      usedRelations.add(relation.symmetric());
+    }
+    return impliedBy(transitiveReduction(knownRelations, usedRelations), usedRelations);
+  }
+
+  private Collection<AbstractSymbolicValueRelation> transitiveReduction(Collection<AbstractSymbolicValueRelation> knownRelations, Set<AbstractSymbolicValueRelation> usedRelations) {
+    boolean changed = false;
+    List<AbstractSymbolicValueRelation> result = new ArrayList<>();
+    for (AbstractSymbolicValueRelation relation : knownRelations) {
+      boolean used = false;
+      if (v1.equals(relation.v1)) {
+        used = result.addAll(relation.combinedRelations(knownRelations, usedRelations));
+      } else if (v1.equals(relation.v2)) {
+        used = result.addAll(relation.symmetric().combinedRelations(knownRelations, usedRelations));
+      }
+      if (used) {
+        changed = true;
+      } else {
+        result.add(relation);
+      }
+    }
+    return changed ? result : Collections.<AbstractSymbolicValueRelation>emptyList();
+  }
+
+  private Collection<AbstractSymbolicValueRelation> combinedRelations(Collection<AbstractSymbolicValueRelation> relations, Set<AbstractSymbolicValueRelation> usedRelations) {
+    List<AbstractSymbolicValueRelation> result = new ArrayList<>();
+    for (AbstractSymbolicValueRelation relation : relations) {
+      AbstractSymbolicValueRelation combined = null;
+      if (v2.equals(relation.v1) && !v1.equals(relation.v2)) {
+        combined = relation.combinedAfter(this);
+      } else if (v2.equals(relation.v2) && !v1.equals(relation.v2)) {
+        combined = relation.symmetric().combinedAfter(this);
+      }
+      if (combined != null && !usedRelations.contains(combined)) {
+        result.add(combined);
+      }
+    }
+    return result;
+  }
+
+  /**
+   * @return a new relation, the negation of the receiver
+   */
+  protected abstract AbstractSymbolicValueRelation inverse();
+
+  /**
+   * @return a new relation, equivalent to the receiver with inverted parameters
+   */
+  protected abstract AbstractSymbolicValueRelation symmetric();
+
+  /**
+   * @param a relation between symbolic values
+   * @return <ul>
+   * <li>TRUE  if the receiver implies that the supplied relation is true</li>
+   * <li>FALSE if the receiver implies that the supplied relation is false</li>
+   * <li>null  otherwise</li>
+   * </ul>
+   */
+  @CheckForNull
+  public Boolean implies(AbstractSymbolicValueRelation relation) {
+    if (v1.equals(relation.v1) && v2.equals(relation.v2)) {
+      return relation.isImpliedBy(this);
+    } else if (v1.equals(relation.v2) && v2.equals(relation.v1)) {
+      return relation.isImpliedBy(symmetric());
+    }
+    return null;
```

Comment:
![MAJOR](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/severity-major.png) Null is returned but a "Boolean" is expected. [![rule](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/rule.png)](https://dory.sonarsource.com/coding_rules#rule_key=squid%3AS2447)